### ISSUE TITLE
How to access the index of element in RuleForEach WithMessage?

### ISSUE THREAD

**Issue by moabtools** [JeremySkinner/FluentValidation#1166]
### System Details

- FluentValidation version: 8.4.0

### Issue Description

How to access the index of element in RuleForEach WithMessage?
As an example:

`
RuleForEach(x => x.SourcePhrases).NotNull().WithMessage("Source phrase in line {Index} cannot be null");
`

Thanks!

**Comment by JeremySkinner** [comment]
It's not exposed directly, I think you'll need to use a custom message builder to extract it and add it into the available placeholders. Give this a try:

```csharp
RuleForEach(x => x.SourcePhrases).NotNull()
.Configure(cfg => {
  cfg.MessageBuilder = context => {
      string propertyNameSoFar = context.ParentContext.PropertyChain.ToString();
      // At this point, propertyNameSoFar should be something like "SourcePhrases[0]"
      // Use a regex to split out the index from the property name. 
     string index = ... // extract the index
     context.MessageFormatter.AppendArgument("Index", index);
     return context.GetDefaultMessage();
  };
})
.WithMessage("Source phrase in line {Index} cannot be null");

```

If that works, you could wrap it in an extension method to clean it up a bit. 

I like the idea though, so can look at adding native support for a future version.

**Comment by JeremySkinner** [comment]
Implemented in 41252af which will be in 8.5
